src/main/java/org/flint/Server.java

```diff
@@ -1,81 +1,95 @@
+import javaslang.Tuple2;
+
+import org.jparsec.error.ParserException;
+
+import org.flint.parse.Http;
+import org.util.FileSystem;
 
 public class Server {
-    List<Route> routes;
+    private Function1<Request, Response> requestHandler;
 
-    public Server(List<Route> routes) {
-        this.routes = routes;
+    public Server(Function1<Request, Response> requestHandler) {
+        this.requestHandler = requestHandler;
     }
 
-    public Response handle(Request request) {
-        Response response;
-        List<Route> resourceMatches = routes.filter((route) -> isUriMatch(route, request));
-
-        if (resourceMatches.isEmpty()) {
-            response = defaultResponse(StatusCode.NOT_FOUND);
-        } else {
-            List<Route> methodMatches = resourceMatches.filter((route) -> isMethodMatch(route, request));
-
-            if (methodMatches.isEmpty()) {
-                response = defaultResponse(StatusCode.METHOD_NOT_ALLOWED);
-                response.setHeader("Allow", buildAllowHeader(resourceMatches));
-            } else {
-                try {
-                    response = methodMatches.head().getController().apply(request);
-                } catch (HttpException he) {
-                    response = defaultResponse(he.getStatusCode());
-                    he.getHeaders().forEach(response::setHeader);
-                } catch (Throwable e) {
-                    response = defaultResponse(StatusCode.INTERNAL_SERVER_ERROR);
+    public void run(int port) throws IOException {
+        try(ServerSocket listener = new ServerSocket(port)) {
+            System.out.println("Server running at http://localhost:" + port);
+            while (true) {
+                try(Socket socket = listener.accept()) {
+                    InputStreamReader reader = new InputStreamReader(socket.getInputStream());
+                    OutputStream os = socket.getOutputStream();
+
+                    handleHttpMessage(reader, os);
                 }
             }
         }
+    }
 
-        if (request.getMethod().equals("HEAD")) {
-            response.removeBody();
+    public void handleHttpMessage(Reader reader, OutputStream os) throws IOException {
+        try {
+            Request request = readRequest(reader);
+            Response response = requestHandler.apply(request);
+            writeResponse(response, os);
+        } catch (ParserException pe) {
+            Response response = Response.create(StatusCode.BAD_REQUEST);
+            writeResponse(response, os);
+        } catch (Throwable t) {
+            Response response = Response.create(StatusCode.INTERNAL_SERVER_ERROR);
+            writeResponse(response, os);
         }
-
-        return response;
     }
 
-    private String buildAllowHeader(List<Route> matches) {
-        return matches
-            .map(Route::getMethod)
-            .reduce((a, b) -> a + "," + b);
-    }
+    private Request readRequest(Reader reader) throws IOException {
```

Comment:
I think this refactor is in the right direction, but I think it can go a bit further. I would extract the `#readRequest` and `#writeResponse` methods into their own classes, that way `Server` can just coordinate each of your components together. It would also reduce the scope of your tests, since your `#readRequest` and `writeResponse` classes could just be injected / mocked at this level.